Assignment 2

fulopbbea opened the pull request:
Sourcecode and test cases for the 2nd assignment
Tree class:
-private TreeNode class implementing recursive tree search method
-addNode method by specifying parent and left/right child
-listAncestors method searching for the node, traversing back on the tree structure to the root using the parent pointer of the TreeNode objects
-lowestCommonAncestor iterating through the lists returned by the listAncestors method finding the first common occurrence of a key

KeyNotFoundException
-thrown whenever a key is not found in the tree

fulopbbea committed:
Assignment 2

Sourcecode and test cases for the 2nd assignment
Tree class:
-private TreeNode class implementing recursive tree search method
-addNode method by specifying parent and left/right child
-listAncestors method searching for the node, traversing back on the tree structure to the root using the parent pointer of the TreeNode objects
-lowestCommonAncestor iterating through the lists returned by the listAncestors method finding the first common occurrence of a key

KeyNotFoundException
-thrown whenever a key is not found in the tree

elifguler96 commented on src/assignment2/KeyNotFoundException.java:
```diff
@@ -0,0 +1,18 @@
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
```
I guess you could just delete this comment

elifguler96 commented on src/assignment2/KeyNotFoundException.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ *
+ * @author bea
+ */
```
You could also delete this one

elifguler96 commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+    }
+    
+    public Optional<Integer> getLowestCommonAncestor(int key1, int key2) {
+        ArrayList<Integer> ancestors1 = getAncestors(key1);
```
We were not allowed to store additional nodes in a data structure, but here you store the ancestors in a list.

brzozan replied:
I really like the solution with lists, but I think Elif is right - could you change implementation (possibly keeping current one too in a separate method) to operate only on the tree?

elifguler96 commented on test/assignment2/TreeTest.java:
```diff
@@ -0,0 +1,109 @@
+        System.out.println("getLowestCommonAncestorNonexistent");
+        int key1 = 14;
+        int key2 = 11;this.tree.getLowestCommonAncestor(key1, key2);
+    }
```
I think you can also include a test for two nodes where one of them is the ancestor of the other.

elifguler96 reviewed (changes requested):
I just left some comments about some possible changes. I also want to say that it would be great if you could write some comments to explain your code :)

brzozan commented on test/assignment2/TreeTest.java:
```diff
@@ -0,0 +1,109 @@
+public class TreeTest {
+    
+    public Tree tree;
+    public Tree nullTree;
```
It does not seem to be needed.

brzozan commented on test/assignment2/TreeTest.java:
```diff
@@ -0,0 +1,109 @@
+    public Tree tree;
+    public Tree nullTree;
+    
+    public TreeTest() {
```
Not needed. Also setUpClass, tearDownClass and tearDown.

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+        return this.root.searchNode(key);
+    }
+    
+    public void addNode(int key, int parentKey, Boolean toLeft) {
```
No need to use boxed type Boolean here. Just boolean - it is more efficient and there is no need to worry about it being null.

brzozan commented on src/assignment2/KeyNotFoundException.java:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @author bea
+ */
+class KeyNotFoundException extends NullPointerException {
```
Extending standard exceptions, like NullPointerException, is not recommended, since they have specific semantics.

If you want to use your own exception class, you could just extend Exception. It would also make it a checked exception rather than runtime one, i.e. methods that can throw it would need to explicitly declare it, and code that calls them would have to explicitly handle it.

If you don't like checked exceptions (even though you should, as explicitly stating custom exceptions that can be thrown has some advantages) you could extend RuntimeException instead.

Another possibility is to just throw IllegalArgumentException instead of custom one.

brzozan commented on src/assignment2/KeyNotFoundException.java:
```diff
@@ -0,0 +1,18 @@
+    
+    public KeyNotFoundException(int key) {
+        String message = "Key " + key + " not found in the tree.";
+        System.err.println(message);
```
Messages are not usually printed from exception's constructor. You should call parent exception's constructor with the message instead, for example [super(message);](https://docs.oracle.com/javase/7/docs/api/java/lang/Exception.html#Exception(java.lang.String)).

Then it will be printed automatically if exception is thrown and not caught, or by explicit code in try..catch block.

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+    private class TreeNode {
+
+        private final int key;
+        private TreeNode left, right, parent;
```
It is a better to declare each field (or variable) separately. It is more verbose, but when reading code a quick glance is enough to scan for what one is looking for rather than needing to parse whole statement.

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+            this.parent = parent;
+        }
+        
+        public TreeNode searchNode(int key) {
```
Minor nitpick: I think that findNode would be a slightly better name.

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+        ArrayList<Integer> ancestors = new ArrayList<>();
+        TreeNode node = this.searchNode(key);
+        if (node != null) {
+            node = node.getParent();
```
This could go inside `for()`: `for(node = node.getParent(); ...)`.

Or even better, to make code more explicit:

`for (TreeNode parent = node.getParent(); parent != null; parent = parent.getParent())`

brzozan commented on test/assignment2/TreeTest.java:
```diff
@@ -0,0 +1,109 @@
+    @Before
+    public void setUp() {
+        this.tree = new Tree(16);
+        tree.addNode(9, 16, Boolean.TRUE);
```
No need to use Boolean.TRUE/FALSE. Just true/false.

brzozan commented on test/assignment2/TreeTest.java:
```diff
@@ -0,0 +1,109 @@
+        ArrayList<Integer> expResult = new ArrayList<>();
+        expResult.add(3);
+        expResult.add(9);
+        expResult.add(16);
```
You can use Arrays.asList(3, 9, 16) instead of manually constructed ArrayList. Additionally you can just use it inline in assertEquals call, without assigning it to separate variable.

The same applies to other test methods.

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+    
+    public ArrayList<Integer> getAncestors(int key) {
```
It is usually better to return most general interface rather than its specific implementation (in this case: `List<Integer>` rather than `ArrayList<Integer>`).

Typically user of the class does not care about exact class, but only about what can be done with it, and if in the future you decide that LinkedList would work better, there is no need to change all the other places where this method is called and assigned to variable of type `ArrayList<>`.

brzozan commented on test/assignment2/TreeTest.java:
```diff
@@ -0,0 +1,109 @@
+    public void testGetLowestCommonAncestorNonexistent() {
+        System.out.println("getLowestCommonAncestorNonexistent");
+        int key1 = 14;
+        int key2 = 11;this.tree.getLowestCommonAncestor(key1, key2);
```
Please break line after ;

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+    
+    public Optional<Integer> getLowestCommonAncestor(int key1, int key2) {
```
I like the idea of using Optional, however I don't think Optional.empty() will ever be returned, since if both keys are present in the tree, they are guaranteed to have common ancestor (possibly root of the tree), but if at least one of them is not in the tree, exception is thrown.

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+            }
+            return ancestors;
+        } else {
+            throw (new KeyNotFoundException(key));
```
No need for brackets around `new KeyNotFoundException(key)`.

brzozan commented on src/assignment2/Tree.java:
```diff
@@ -0,0 +1,120 @@
+                parent.setRightChild(new TreeNode(key, parent));
+            }
+        } else {
+            throw (new KeyNotFoundException(parentKey));
```
No need for brackets around `new KeyNotFoundException(key)`.

brzozan reviewed (changes requested):
Very nice and well tested solution!

I added a few, mostly minor comments.

fulopbbea committed:
Doc, printAncestors, new test cases added

-added documentation for relevant methods
-added printAncestor method to print the ancestors of a node without any additional data structs and test cases for it
-replaced custom Exception with IllegalArgumentException
-tidied up the code

fulopbbea committed:
deleted KeyNotFoundException

-not needed anymore; replaced by IllegalArgumentException

brzozan reviewed (approved):
Thanks!
